lib/auth/cryptosign.ts

```diff
@@ -0,0 +1,84 @@
+import * as nacl from 'tweetnacl';
+
+import {Authenticate, AuthenticateFields} from "../messages/authenticate";
+import Challenge from "../messages/challenge";
+import Authenticator from "./authenticator";
+
+export class CryptoSignAuthenticator implements Authenticator {
+    static TYPE = "cryptosign"
+
+    constructor(
+        private readonly _authid: string,
+        private readonly _privateKey: string,
+        private readonly _authExtra: { [key: string]: any }
+    ) {}
+
+    get authMethod(): string {
+        return CryptoSignAuthenticator.TYPE;
+    }
+
+    get authID(): string {
+        return this._authid;
+    }
+
+    get authExtra(): object {
+        return this._authExtra;
+    }
+
+    authenticate(challenge: Challenge): Authenticate {
+        if (!challenge.extra.hasOwnProperty("challenge")) {
+            throw new Error("challenge string missing in extra");
+        }
+        const challengeHex: string = challenge.extra["challenge"];
+
+        const signed: string = signCryptoSignChallenge(challengeHex, this._privateKey);
+
+        return new Authenticate(new AuthenticateFields(signed, {}));
+    }
+}
+
+export function hexToUint8Array(hex: string): Uint8Array {
+  return new Uint8Array(Buffer.from(hex, 'hex'));
+}
+
+export function uint8ArrayToString(bytes: Uint8Array): string {
+    return Array.from(bytes).map(byte => byte.toString(16).padStart(2, '0')).join('');
+}
+
+
+export function signCryptoSignChallenge(challenge: string, privateKeyHex: string): string {
+    let privateKey: Uint8Array;
+    if (privateKeyHex.length == 64) {
+        const key = hexToUint8Array(privateKeyHex);
+        const keyPair = nacl.sign.keyPair.fromSeed(hexToUint8Array(privateKeyHex));
+        const publicKey = keyPair.publicKey;
+
+        // combine private & public keys
+        privateKey = new Uint8Array(64);
+        privateKey.set(key);
+        privateKey.set(publicKey, 32);
+    } else {
+        throw new Error("invalid private key length")
+    }
+
+    const rawChallenge = hexToUint8Array(challenge);
+    const signature = nacl.sign(rawChallenge, privateKey);
+
+    return uint8ArrayToString(signature);
```

Comment:
the signature needs to append the challenge at the end

Reply:
`nacl.sign()` already adds the challenge to the signature